src/backend.py

```diff
@@ -0,0 +1,726 @@
+                return func(request)
+            finally:
+                running = False
+
+        return wrapped_route
+    return decorator
+
+def four_oh_four(request):
+    print("--- 404 ---")
+    print(request)
+    print(request.query)
+    print()
+    return "Not found", 404
+
+def redirect(request):
+    from phew.server import redirect
+    #TODO check that the route is a valid one for the server, else 404
+    return redirect(f"/index.html", status=303)
+#
+# Authentication
+#
+def hmac_sha256(key, message):
+    """
+    Compute HMAC-SHA256 using the given key and message.
+    key and message should be bytes.
+    """
+    BLOCK_SIZE = 64  # block size for SHA-256
+
+    # If key longer than block size, hash it
+    if len(key) > BLOCK_SIZE:
+        h = hashlib_sha256()
+        h.update(key)
+        key = h.digest()
+
+    # Pad the key
+    if len(key) < BLOCK_SIZE:
+        key = key + b'\x00' * (BLOCK_SIZE - len(key))
+
+    # Create inner and outer pads
+    o_key_pad = bytes([b ^ 0x5c for b in key])
+    i_key_pad = bytes([b ^ 0x36 for b in key])
+
+    # Inner hash
+    h_inner = hashlib_sha256()
+    h_inner.update(i_key_pad + message)
+    inner_hash = h_inner.digest()
+
+    # Outer hash
+    h_outer = hashlib_sha256()
+    h_outer.update(o_key_pad + inner_hash)
+    return h_outer.digest()
+
+def require_auth(handler):
+    '''Decorator to require authentication using HMAC-SHA256'''
+    def auth_wrapper(request, *args, **kwargs):
+        global current_challenge, challenge_timestamp
+
+        if not current_challenge or not challenge_timestamp:
+            msg = json_dumps({"error": "Challenge missing"}), 401, 'application/json'
+            print(msg)
+            return msg
+
+        if (time() - challenge_timestamp) > CHALLENGE_EXPIRATION_SECONDS:
+            msg = json_dumps({"error": "Challenge expired"}), 401, 'application/json'
+            print(msg)
+            return msg
+
+        # Get the HMAC from the request headers
+        client_hmac = request.headers.get('X-Auth-HMAC') or request.headers.get('x-auth-hmac')
+        if not client_hmac:
+            msg = json_dumps({"error": "Missing authentication HMAC"}), 401, 'application/json'
+            print(msg)
+            print(request.headers)
+            return msg
+
+        # Retrieve stored password (the key)
+        credentials = ds_read_record("configuration", 0)
+        stored_password = credentials["Gpassword"].encode('utf-8')
+
+        # Construct the message string
+        path = request.path
+        query_string = request.query_string or ""
+        body_str = request.raw_data or ""
+        message_str = current_challenge + path + query_string + body_str
+        message_bytes = message_str.encode('utf-8')
+
+        # Compute expected HMAC
+        expected_digest = hmac_sha256(stored_password, message_bytes)
+        expected_hmac = hexlify(expected_digest).decode()
+
+        if client_hmac == expected_hmac:
```

Comment:
This is how we authenticate requests. We include the request query/data in the hash so a man in the middle attack would theoretically be unable to alter what the command does. The current challange changes every time too, so they can't replay a command like reseting the game or something.